### PR TITLE
kerberos: authentication between client and broker

### LINKED ISSUE #3652: Kerberos authenticate between client and broker
**Is your feature request related to a problem? Please describe.**
Kerberos authenticate between client and broker.

**Additional context**
This is a sub task of issue #3491

### PR DESCRIPTION
**Pull request by jiazhai**
Fixes #3652

**Motivation**
Currently both Zookeeper and BookKeeper could be secured by using Kerberos, It would be good to support Kerberos in Pulsar Broker and Client.
This is the sub-task for issue in #3491 to support Kerberos in Pulsar Broker and Client.
Will add proxy and web resource support in following prs.
The Kerberos authentication is similar to that in Zookeeper and BookKeeper, which leverage SASL and GSSAPI, so reused some of the code there.
PR #3658 is the first version of PR before #3677 .

**Changes**
provide both client and broker side support for authentication api;
add unit test.

### CODE REVIEW

**Comment by jiazhai** [comment]
@eolivelli  Thanks for the comments in #3658. This is the second version.

**Comment by eolivelli** [comment]
@jiazhai I will take a look

File: conf/broker.conf [hunk 1]
```diff
@@ -340,6 +340,22 @@ tokenSecretKey=
 # tokenPublicKey=file:///my/public.key
 tokenPublicKey=
 
+### --- SASL Authentication Provider --- ###
+
+# Whether Use SASL Authentication or not.
+# TODO: used to bypass web resource check. will remove it after implementation the support.
```
Comment (eolivelli) [inline conf/broker.conf]: TODO ?

Comment (jiazhai) [reply conf/broker.conf]: Thanks @eolivelli , yes, there is a planed work. 
current step is for client <-> broker. will add the web-resource/http support in the TODO issue.

Comment (jiazhai) [reply conf/broker.conf]: by this parameter we could bypass the web-resource/http checking temporary.

**Pull request merged**